src/scanner/mod.rs

```diff
@@ -100,41 +103,59 @@ impl Scanner {
     ///
     /// Note: `self` must contain `self.ip`.
     async fn scan_socket(&self, socket: SocketAddr) -> io::Result<SocketAddr> {
-        match self.connect(socket).await {
-            Ok(x) => {
-                debug!(
-                    "Connection was successful, shutting down stream {}",
-                    &socket
-                );
-                match x.shutdown(Shutdown::Both) {
-                    Err(e) => debug!("Shutdown stream error {}", &e),
-                    _ => {}
-                }
-                if !self.greppable {
-                    if self.accessible {
-                        println!("Open {}", socket.to_string());
-                    } else {
-                        println!("Open {}", socket.to_string().purple());
+        let mut ret = Err(io::Error::new(
+            io::ErrorKind::Other,
+            format!("Did not connect to {}", &socket),
+        ));
+
+        for nr_try in 0..self.tries {
+            debug!("Try number: {}", nr_try + 1);
+
+            ret = match self.connect(socket).await {
+                Ok(x) => {
+                    debug!(
+                        "Connection was successful, shutting down stream {}",
+                        &socket
+                    );
+                    match x.shutdown(Shutdown::Both) {
+                        Err(e) => debug!("Shutdown stream error {}", &e),
+                        _ => {}
                     }
+                    if !self.greppable {
+                        if self.accessible {
+                            println!("Open {}", socket.to_string());
+                        } else {
+                            println!("Open {}", socket.to_string().purple());
+                        }
+                    }
+
+                    Ok(socket)
                 }
+                Err(e) => match e.kind() {
+                    ErrorKind::Other => {
+                        if e.to_string().contains("No route to host")
+                            || e.to_string().contains("Network is unreachable")
+                        {
+                            debug!("Socket connect error: {} {}", &e.to_string(), &socket);
+                            Err(io::Error::new(io::ErrorKind::Other, e.to_string()))
+                        } else {
+                            debug!("Socket connect error: {} {}", &e.to_string(), &socket);
+                            panic!("Too many open files. Please reduce batch size. The default is 5000. Try -b 2500.");
+                        }
+                    }
+                    _ => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
```

Comment:
The `to_string` method is called many times in this scope. I am not sure if the compiler can optimize this away and only perform one allocation, but I think it would be nicer if you stored the `String` in a local variable and used that.

Reply:
I did not see this problem when I was coding, I corrected it on my side.